### PR TITLE
Minh -> Master

### PR DESCRIPTION
**Pull request by minheragon12345**

### CODE REVIEW

File: src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        return Monster.createMonsterAttributes()
+                .add(Attributes.MAX_HEALTH, 15.0D)
+                .add(Attributes.ATTACK_DAMAGE, 4.0f)
+                .add(Attributes.ARMOR, 1.0f)
+                .add(Attributes.KNOCKBACK_RESISTANCE, 0.01f)
+                .add(Attributes.FOLLOW_RANGE, 64.0f)
+                .add(Attributes.MOVEMENT_SPEED, 1.5F).build();
+    }
+
+    protected void registerGoals() {
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java:52]: Missing override annotation

File: src/main/java/com/github/manasmods/tensura/event/ModEventBusEvents.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import com.github.manasmods.tensura.Tensura;
+import com.github.manasmods.tensura.entity.ModEntityTypes;
+import com.github.manasmods.tensura.entity.custom.GiantSalmonEntity;
+import net.minecraftforge.event.entity.EntityAttributeCreationEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+
+
+@Mod.EventBusSubscriber(modid = Tensura.MOD_ID, bus = Mod.EventBusSubscriber.Bus.MOD)
+public class ModEventBusEvents {
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/event/ModEventBusEvents.java:12]: Rename the class to `TensuraEntityHandler`

File: src/main/java/com/github/manasmods/tensura/event/ModEventBusEvents.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    @SubscribeEvent
+    public static void entityAttributeEvent(EntityAttributeCreationEvent event) {
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/event/ModEventBusEvents.java:15]: remove annotation and add this method as listener to the modEventBus using `modEventBus.addListener(<class name>:<method name>)` in the Tenura mod constructor

File: src/main/java/com/github/manasmods/tensura/item/TensuraItems.java [hunk 4]
```diff
@@ -1,8 +1,13 @@
 package com.github.manasmods.tensura.item;
 
 import com.github.manasmods.tensura.Tensura;
+import com.github.manasmods.tensura.entity.ModEntityTypes;
+import net.minecraft.world.item.CreativeModeTab;
 import net.minecraft.world.item.Item;
+import net.minecraftforge.common.ForgeSpawnEggItem;
+import net.minecraftforge.event.RegistryEvent;
 import net.minecraftforge.registries.ObjectHolder;
+import net.minecraftforge.registries.RegistryObject;
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/item/TensuraItems.java:10]: unused import?

File: src/main/java/com/github/manasmods/tensura/world/ModWorldEvents.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import net.minecraftforge.fml.common.Mod;
+
+@Mod.EventBusSubscriber(modid = Tensura.MOD_ID)
+public class ModWorldEvents {
+    @SubscribeEvent
+    public static void biomeLoadingEvent(final BiomeLoadingEvent event) {
+
+        ModEntityGeneration.onEntitySpawn(event);
+    }
+}
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/world/ModWorldEvents.java:16]: The whole class can be removed

File: src/main/java/com/github/manasmods/tensura/world/entity/ModEntityGeneration.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.level.biome.Biome;
+import net.minecraft.world.level.biome.MobSpawnSettings;
+import net.minecraftforge.event.world.BiomeLoadingEvent;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class ModEntityGeneration {
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/world/entity/ModEntityGeneration.java:13]: Make this class an EventBusSubscriber

File: src/main/java/com/github/manasmods/tensura/world/entity/ModEntityGeneration.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    public static void onEntitySpawn(final BiomeLoadingEvent event) {
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/world/entity/ModEntityGeneration.java:14]: Subscribe to the event

**Review by Charismara — CHANGES_REQUESTED** [review]
Just a few changes. Overall good job!

File: src/main/java/com/github/manasmods/tensura/entity/ai/AIMoveTowardsTargetGoal.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            if (!this.attacker.getNavigation().moveTo(livingentity, this.speedTowardsTarget)) {
+                this.delayCounter += 15;
+            }
+        }
+    }
+
+    protected double getAttackReachSqr(LivingEntity attackTarget) {
+        return this.attacker.getBbWidth() * 2.0D * this.attacker.getBbWidth() * 2.0D + attackTarget.getBbWidth();
+    }
+}
```
[the remark this answers is not in the document]
Comment (minheragon12345) [reply src/main/java/com/github/manasmods/tensura/entity/ai/AIMoveTowardsTargetGoal.java:112]: Better AI for swimming mobs.

File: src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                    speed = 1.5F;
+                    GiantSalmonEntity.this.lookAt(target, 70, 70);
+                }
+            }
+            if (target instanceof Drowned || target instanceof Player) {
+                speed = 2F;
+            }
+            GiantSalmonEntity.this.getNavigation().moveTo(target, speed);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (minheragon12345) [reply src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java:184]: This new AI can change attack reach and change speed based on distance.

[the remark this answers is not in the document]
Comment (Charismara) [reply src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java:184]: fair enough. exctract it in a new class please. We might reuse it for other entities :)

File: src/main/java/com/github/manasmods/tensura/entity/ai/AIMoveTowardsTargetGoal.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Charismara) [reply src/main/java/com/github/manasmods/tensura/entity/ai/AIMoveTowardsTargetGoal.java:112]: Okay then rename the class please to `SwimTowardsTargetGoal`

File: src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (minheragon12345) [reply src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java:184]: we can change the speed and reach in that entities and add more conditions to change speed if needed, every entity is different so I cant make it a new class xD

File: src/main/java/com/github/manasmods/tensura/entity/ai/AIMoveTowardsTargetGoal.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (minheragon12345) [reply src/main/java/com/github/manasmods/tensura/entity/ai/AIMoveTowardsTargetGoal.java:112]: Tho it still works with Normal Land Creature lol

File: src/main/java/com/github/manasmods/tensura/Tensura.java [hunk 10]
```diff
@@ -55,7 +52,7 @@ private void setup(final FMLCommonSetupEvent event) {
         //Register our Network Channel
         event.enqueueWork(TensuraNetwork::register);
 
-        SpawnPlacements.register(ModEntityTypes.GIANT_SALMON.get(),
+        SpawnPlacements.register(TensuraEntityTypes.GIANT_SALMON.get(),
                 SpawnPlacements.Type.IN_WATER,
                 Heightmap.Types.MOTION_BLOCKING_NO_LEAVES,
                 WaterAnimal::checkSurfaceWaterAnimalSpawnRules);
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/Tensura.java:58]: Create a new Method called `registerEntityPlacements` in the `TensuraEntityHandler` and call it in the `Tensura#setup` Method using `event.enqueueWork(TensuraEntityHandler::registerEntityPlacements)`

File: src/main/java/com/github/manasmods/tensura/world/entity/TensuraEntityGeneration.java [hunk 11]
```diff
@@ -1,18 +1,23 @@
...
 import net.minecraft.world.level.biome.MobSpawnSettings;
 import net.minecraftforge.event.world.BiomeLoadingEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
 
 import java.util.Arrays;
 import java.util.List;
 
-public class ModEntityGeneration {
+@Mod.EventBusSubscriber(modid = Tensura.MOD_ID)
```
Comment (Charismara) [inline src/main/java/com/github/manasmods/tensura/world/entity/TensuraEntityGeneration.java:16]: set used bus to FORGE

**Review by Charismara — CHANGES_REQUESTED** [review]
Just 2 more things :)

File: src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Charismara) [reply src/main/java/com/github/manasmods/tensura/entity/custom/GiantSalmonEntity.java:185]: Line 169 seems unused btw. remove it please :)
Why shouldn't it possible to reuse this Goal atleast for other Fish/Swimming Entities?

**Review by Charismara — CHANGES_REQUESTED** [review]
Gonna do Goal thing on my own

**Review by Charismara — APPROVED** [review]
Let's wait for CI/CD Pipeline

**Pull request merged**